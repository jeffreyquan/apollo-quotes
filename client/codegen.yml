schema: http://localhost:5000/graphql
generates:
  ./types.ts:
    plugins:
      - typescript
