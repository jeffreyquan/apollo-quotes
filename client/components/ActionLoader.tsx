import React from "react";
import ClipLoader from "react-spinners/ClipLoader";
import { ThemeConsumer } from "styled-components";

export const ActionLoader: React.FC = () => {
  return (
    <ThemeConsumer>
      {(theme) => <ClipLoader color={theme.actionloader} size={20} />}
    </ThemeConsumer>
  );
};
