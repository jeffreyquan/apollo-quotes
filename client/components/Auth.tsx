import { createContext, useState, useEffect } from "react";
import { useQuery, gql } from "@apollo/client";

const CURRENT_USER_QUERY = gql`
  query {
    userProfile {
      id
      name
      username
      email
      role
    }
  }
`;

export const AuthContext = createContext({
  checkingAuth: true,
  user: null,
  setUser: null,
});

export const AuthProvider = ({ children }) => {
  const [checkingAuth, setCheckingAuth] = useState(true);
  const [user, setUser] = useState(null);

  const { data, loading, error } = useQuery(CURRENT_USER_QUERY);

  useEffect(() => {
    if (typeof data !== undefined) {
      setCheckingAuth(false);
      if (data) {
        setUser(data.userProfile);
      }
    }
  }, [data]);

  const value = { checkingAuth, user, setUser };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};
