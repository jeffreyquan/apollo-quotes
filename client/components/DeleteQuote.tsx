import React from "react";
import { gql, useMutation } from "@apollo/client";
import { BsExclamationCircle } from "react-icons/bs";
import { useRouter } from "next/router";
import { ActionLoader } from "./ActionLoader";

const DELETE_QUOTE_MUTATION = gql`
  mutation DELETE_QUOTE_MUTATION($id: ID!) {
    deleteQuote(id: $id) {
      id
    }
  }
`;

interface DeleteQuoteProps {
  id: string;
}

export const DeleteQuote: React.FC<DeleteQuoteProps> = ({ id, children }) => {
  const [deleteQuote, { error, loading }] = useMutation(DELETE_QUOTE_MUTATION, {
    variables: { id },
    update(cache, { data: { deleteQuote } }) {
      cache.modify({
        fields: {
          quotes(
            existing = {
              __typename: "QuotesConnection",
              totalCount: 0,
              pageInfo: {
                endCursor: null,
                hasMore: true,
              },
              quotes: [],
            },
            { readField }
          ) {
            const updatedQuotes = existing.quotes.filter(
              (quoteRef) => readField("id", quoteRef) !== deleteQuote.id
            );
            return {
              ...existing,
              quotes: [...updatedQuotes],
            };
          },
        },
      });
    },
  });

  if (loading) return <ActionLoader />;
  if (error) return <BsExclamationCircle />;

  const router = useRouter();

  const handleClick = async (e) => {
    e.preventDefault();
    try {
      await deleteQuote();
      router.push({
        pathname: "/quotes",
        query: { delete: "success" },
      });
    } catch (err) {
      return <BsExclamationCircle />;
    }
  };

  return (
    <div
      data-testid="deleteButton"
      onClick={(e) => {
        if (confirm("Please confirm you want to delete this quote.")) {
          handleClick(e);
        }
      }}
    >
      {children}
    </div>
  );
};
