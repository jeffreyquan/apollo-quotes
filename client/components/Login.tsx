import { useState, useEffect, useContext } from "react";
import { useRouter } from "next/router";
import { gql, useMutation } from "@apollo/client";
import { AuthContext } from "./Auth";
import { useForm } from "../lib/useForm";
import { Form } from "../styles/Form";
import { FormContainer } from "../styles/FormContainer";
import { FormTitle } from "../styles/FormTitle";

const LOGIN_MUTATION = gql`
  mutation LOGIN_MUTATION($email: String!, $password: String!) {
    login(email: $email, password: $password) {
      id
      name
      email
      username
    }
  }
`;

export const Login = () => {
  const { inputs, handleChange, resetForm } = useForm({
    email: "",
    password: "",
  });

  const { email, password } = inputs;

  const [loadingPage, setLoadingPage] = useState(true);

  let { user, setUser } = useContext(AuthContext);

  const router = useRouter();

  useEffect(() => {
    if (user) {
      if (router.query.redirect) {
        router.back();
      } else {
        router.push("/");
      }
    } else {
      setLoadingPage(false);
    }
  }, [user]);

  const [login, { error, loading }] = useMutation(LOGIN_MUTATION, {
    variables: inputs,
  });

  // TODO: create error component to display any login errors
  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const res = await login();
      setUser(res.data.login);
    } catch (err) {
      console.log(err);
    }
  };

  return !loadingPage ? (
    <FormContainer>
      <Form onSubmit={handleSubmit}>
        <FormTitle>Login</FormTitle>
        <fieldset disabled={loading} aria-busy={loading}>
          <label htmlFor="email">
            Email
            <input
              type="email"
              name="email"
              placeholder="Email"
              value={email}
              onChange={handleChange}
              autoComplete="email"
            />
          </label>
          <label htmlFor="password">
            Password
            <input
              type="password"
              name="password"
              placeholder="Password"
              value={password}
              onChange={handleChange}
              autoComplete="current-password"
            />
          </label>
          <input type="submit" value="Sign In" />
        </fieldset>
      </Form>
    </FormContainer>
  ) : (
    <div>Loading....</div>
  );
};
