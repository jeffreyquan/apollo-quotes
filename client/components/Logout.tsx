import { useContext } from "react";
import { gql, useMutation } from "@apollo/client";
import { AuthContext } from "./Auth";

const LOGOUT_MUTATION = gql`
  mutation LOGOUT_MUTATION {
    logout {
      message
    }
  }
`;

export const Logout = () => {
  const [logout] = useMutation(LOGOUT_MUTATION);

  let { setUser } = useContext(AuthContext);

  const logoutUser = async () => {
    const res = await logout();
    if (res.data.logout.message) {
      setUser(null);
    }
  };

  return (
    <button type="button" onClick={logoutUser}>
      Logout
    </button>
  );
};
