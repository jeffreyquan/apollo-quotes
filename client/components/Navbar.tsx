import Link from "next/link";
import styled from "styled-components";
import { AiFillHome } from "react-icons/ai";
import { IoMdLogIn, IoMdLogOut } from "react-icons/io";
import {
  MdInsertDriveFile,
  MdAssignmentInd,
  MdNoteAdd,
  MdTurnedIn,
  MdAddBox,
  MdCreateNewFolder,
} from "react-icons/md";
import { RiHomeHeartLine } from "react-icons/ri";
import { AuthContext } from "./Auth";
import { Logout } from "./Logout";

const IconStyles = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  line-height: 1.4;
  font-size: 1.2rem;

  svg {
    font-size: 3rem;
  }
`;

const NavbarStyles = styled.nav`
  @media only screen and (max-width: ${(props) => props.theme.smallbp}) {
    display: flex;
    background-color: ${(props) => props.theme.white};
    justify-content: center;
    position: fixed;
    align-items: center;
    bottom: 0;
    width: 100%;
    height: 9rem;
  }
  ul {
    display: flex;

    a {
      color: ${(props) => props.theme.headline2};
      padding: 2rem 1rem;
    }
  }
`;

const IconContainer = ({ children }) => {
  return <IconStyles>{children}</IconStyles>;
};

export const Navbar = () => {
  return (
    <AuthContext.Consumer>
      {({ user }) => (
        <NavbarStyles>
          <ul>
            <Link href="/">
              <a>
                <IconContainer>
                  <AiFillHome />
                  <p>Quotes</p>
                </IconContainer>
              </a>
            </Link>
            {user && (
              <>
                <Link
                  href="/user/[...slug]"
                  as={`/user/${user.username}/quotes`}
                >
                  <a>
                    <IconContainer>
                      <MdAssignmentInd />
                      My Quotes
                    </IconContainer>
                  </a>
                </Link>
                <Link href="/quotes/new">
                  <a>
                    <IconContainer>
                      <MdNoteAdd />
                      New
                    </IconContainer>
                  </a>
                </Link>
              </>
            )}
            {!user ? (
              <Link href="/signin">
                <a>
                  <IconContainer>
                    <IoMdLogIn />
                    <p>Sign In</p>
                  </IconContainer>
                </a>
              </Link>
            ) : (
              <Logout>
                <IconContainer>
                  <IoMdLogOut />
                  Log out
                </IconContainer>
              </Logout>
            )}
          </ul>
        </NavbarStyles>
      )}
    </AuthContext.Consumer>
  );
};
