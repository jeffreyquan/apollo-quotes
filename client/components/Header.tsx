import Link from "next/link";
import { useState, useEffect } from "react";
import styled from "styled-components";
import { IoIosRocket } from "react-icons/io";
import { Navbar } from "./Navbar";

const Logo = styled.div`
  font-size: 3rem;
  margin-right: auto;
  a {
    display: flex;
    align-items: center;
  }
`;

const StyledHeader = styled.header`
  position: sticky;
  background-color: ${(props) =>
    props.sticky ? props.theme.white : "transparent"};
  box-shadow: ${(props) =>
    props.sticky ? `0 0.2rem 0.5rem rgba(0,0,0,0.5)` : ""};
  top: 0;
  display: flex;
  align-items: center;
  transition: all 0.3s ease-in-out;
  line-height: ${(props) => (props.sticky ? "3rem" : "5rem")};
  max-height: 9rem;
`;

export const Header = () => {
  const [sticky, setSticky] = useState(false);

  useEffect(() => {
    window.addEventListener("scroll", () => {
      if (window.scrollY > 0) {
        setSticky(true);
      } else {
        setSticky(false);
      }
    });
  }, []);

  return (
    <StyledHeader sticky={sticky}>
      <Logo>
        <Link href="/">
          <a>
            <IoIosRocket />
            Apollo Quotes
          </a>
        </Link>
      </Logo>
      <Navbar />
    </StyledHeader>
  );
};
