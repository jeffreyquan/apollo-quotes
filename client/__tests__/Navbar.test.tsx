import { render, screen, waitFor } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import { Navbar } from "../components/Navbar";
import { AuthProvider } from "../components/Auth";
import { CURRENT_USER_QUERY } from "../components/Auth";

let mutationWasCalled = false;

const userSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: () => {
      mutationWasCalled = true;
      return {
        data: {
          userProfile: {
            id: "abcd",
            name: "Jeffrey",
            username: "jeffrey",
            email: "jeffrey@example.com",
            role: "ADMIN",
            __typename: "User",
          },
        },
      };
    },
  },
];

const userNotSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        userProfile: null,
      },
    },
  },
];

describe("Navbar", () => {
  it("shows home and sign in icons when not signed in", async () => {
    const { getByRole, queryByRole } = render(
      <MockedProvider mocks={userNotSignedInMocks}>
        <Navbar />
      </MockedProvider>
    );

    expect(getByRole("link", { name: "Quotes" })).toBeInTheDocument();
    expect(getByRole("link", { name: "Sign In" })).toBeInTheDocument();
    expect(queryByRole("link", { name: "My Quotes" })).not.toBeInTheDocument();
    expect(queryByRole("link", { name: "Likes" })).not.toBeInTheDocument();
    expect(queryByRole("link", { name: "New" })).not.toBeInTheDocument();
    expect(queryByRole("link", { name: "Log out" })).not.toBeInTheDocument();
  });

  it("shows home, my quotes, likes, new and logout icons whened not signed in", async () => {
    const { getByRole, queryByRole } = render(
      <MockedProvider mocks={userSignedInMocks}>
        <AuthProvider>
          <Navbar />
        </AuthProvider>
      </MockedProvider>
    );

    await waitFor(() => {
      expect(mutationWasCalled).toBe(true);
    });

    expect(getByRole("link", { name: "Quotes" })).toBeInTheDocument();
    await waitFor(() =>
      expect(queryByRole("link", { name: "Sign In" })).not.toBeInTheDocument()
    );
    expect(getByRole("link", { name: "My Quotes" })).toBeInTheDocument();
    expect(getByRole("link", { name: "Likes" })).toBeInTheDocument();
    expect(getByRole("link", { name: "New" })).toBeInTheDocument();
    expect(screen.getByText("Log out")).toBeInTheDocument();
  });
});
