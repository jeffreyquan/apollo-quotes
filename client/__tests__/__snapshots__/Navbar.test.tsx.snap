// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar shows home and sign in icons when not signed in 1`] = `
<div>
  <nav
    class="Navbar__NavbarStyles-sc-188tfqd-1 ijharB"
  >
    <ul>
      <a>
        <div
          class="Navbar__IconStyles-sc-188tfqd-0 ctkHDn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8 3.293l6 6V13.5a1.5 1.5 0 01-1.5 1.5h-9A1.5 1.5 0 012 13.5V9.293l6-6zm5-.793V6l-2-2V2.5a.5.5 0 01.5-.5h1a.5.5 0 01.5.5z"
              fill-rule="evenodd"
            />
            <path
              clip-rule="evenodd"
              d="M7.293 1.5a1 1 0 011.414 0l6.647 6.646a.5.5 0 01-.708.708L8 2.207 1.354 8.854a.5.5 0 11-.708-.708L7.293 1.5z"
              fill-rule="evenodd"
            />
          </svg>
          <p>
            Quotes
          </p>
        </div>
      </a>
      <a>
        <div
          class="Navbar__IconStyles-sc-188tfqd-0 ctkHDn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M10 11H4V3a1 1 0 0 1 1-1h14a1 1 0 0 1 1 1v18a1 1 0 0 1-1 1H5a1 1 0 0 1-1-1v-8h6v3l5-4-5-4v3z"
              />
            </g>
          </svg>
          <p>
            Sign In
          </p>
        </div>
      </a>
    </ul>
  </nav>
</div>
`;

exports[`Navbar shows home, my quotes, likes, new and logout icons whened not signed in 1`] = `
<div>
  <nav
    class="Navbar__NavbarStyles-sc-188tfqd-1 ijharB"
  >
    <ul>
      <a>
        <div
          class="Navbar__IconStyles-sc-188tfqd-0 ctkHDn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8 3.293l6 6V13.5a1.5 1.5 0 01-1.5 1.5h-9A1.5 1.5 0 012 13.5V9.293l6-6zm5-.793V6l-2-2V2.5a.5.5 0 01.5-.5h1a.5.5 0 01.5.5z"
              fill-rule="evenodd"
            />
            <path
              clip-rule="evenodd"
              d="M7.293 1.5a1 1 0 011.414 0l6.647 6.646a.5.5 0 01-.708.708L8 2.207 1.354 8.854a.5.5 0 11-.708-.708L7.293 1.5z"
              fill-rule="evenodd"
            />
          </svg>
          <p>
            Quotes
          </p>
        </div>
      </a>
      <a>
        <div
          class="Navbar__IconStyles-sc-188tfqd-0 ctkHDn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="10"
              rx="1.5"
              transform="matrix(1 0 0 -1 0 14.5)"
              width="16"
            />
            <path
              clip-rule="evenodd"
              d="M2 3a.5.5 0 00.5.5h11a.5.5 0 000-1h-11A.5.5 0 002 3zm2-2a.5.5 0 00.5.5h7a.5.5 0 000-1h-7A.5.5 0 004 1z"
              fill-rule="evenodd"
            />
          </svg>
          <p>
            My Quotes
          </p>
        </div>
      </a>
      <a>
        <div
          class="Navbar__IconStyles-sc-188tfqd-0 ctkHDn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8 1.314C12.438-3.248 23.534 4.735 8 15-7.534 4.736 3.562-3.248 8 1.314z"
              fill-rule="evenodd"
            />
          </svg>
          <p>
            Likes
          </p>
        </div>
      </a>
      <a>
        <div
          class="Navbar__IconStyles-sc-188tfqd-0 ctkHDn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M16 8A8 8 0 110 8a8 8 0 0116 0zM8.5 4a.5.5 0 00-1 0v3.5H4a.5.5 0 000 1h3.5V12a.5.5 0 001 0V8.5H12a.5.5 0 000-1H8.5V4z"
              fill-rule="evenodd"
            />
          </svg>
          <p>
            New
          </p>
        </div>
      </a>
      <div
        class="Logout__LogoutButton-sc-1yl6k4f-0 dyZEYj"
        type="button"
      >
        <div
          class="Navbar__IconStyles-sc-188tfqd-0 ctkHDn"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M5 22a1 1 0 0 1-1-1V3a1 1 0 0 1 1-1h14a1 1 0 0 1 1 1v18a1 1 0 0 1-1 1H5zm10-6l5-4-5-4v3H9v2h6v3z"
              />
            </g>
          </svg>
          <p>
            Log out
          </p>
        </div>
      </div>
    </ul>
  </nav>
</div>
`;
